core/src/cluster_info.rs

```diff
@@ -913,29 +912,31 @@ impl ClusterInfo {
     /// each Vec is no larger than `PROTOCOL_PAYLOAD_SIZE`
     /// Note: some messages cannot be contained within that size so in the worst case this returns
     /// N nested Vecs with 1 item each.
-    fn split_gossip_messages(mut msgs: Vec<CrdsValue>) -> Vec<Vec<CrdsValue>> {
+    fn split_gossip_messages(msgs: Vec<CrdsValue>) -> Vec<Vec<CrdsValue>> {
         let mut messages = vec![];
-        while !msgs.is_empty() {
-            let mut payload = vec![];
-            let mut size = serialized_size(&payload).expect("Couldn't check size");
-            while let Some(msg) = msgs.pop() {
-                let msg_size = msg.size();
-                if size + msg_size > MAX_PROTOCOL_PAYLOAD_SIZE as u64 {
-                    if msg_size < MAX_PROTOCOL_PAYLOAD_SIZE as u64 {
-                        msgs.push(msg);
-                    } else {
-                        debug!(
-                            "dropping message larger than the maximum payload size {:?}",
-                            msg
-                        );
-                    }
-                    break;
+        let mut payload = vec![];
+        let mut size = 0;
+        let max =
+            MAX_PROTOCOL_PAYLOAD_SIZE - serialized_size(&payload).expect("Couldn't check size");
+        for msg in msgs {
+            let msg_size = msg.size();
+            if size + msg_size > max as u64 {
+                if msg_size < max as u64 {
+                    messages.push(payload);
+                    payload = vec![];
+                    size = 0;
+                } else {
+                    debug!(
+                        "dropping message larger than the maximum payload size {:?}",
+                        msg
+                    );
                 }
-                size += msg_size;
-                payload.push(msg);
+                continue;
```

Comment:
This is incorrect, it needs to do what my PR is doing. the "msg" is dropped here because of the continue.

Reply:
land your fix, ill rebase